replication-api/src/main/java/com/connexta/replication/queue/Task.java

```diff
@@ -0,0 +1,154 @@
+   * {@link #putForRetry(PriorityLevel)} has been called on it.
+   *
+   * @return <code>true</code> if the current thread/worker owns the lock to this task; <code>false
+   *     </code> otherwise
+   */
+  public boolean isLocked();
+
+  /**
+   * Attempts to unlock this task. An unlock task will be picked up by the next worker attempting to
+   * poll for a task from the associated queue.
+   *
+   * @throws IllegalStateException if the current thread/worker doesn't own the lock to the task or
+   *     if the task is not current locked
+   * @throws InterruptedException if the thread was interrupted while attempting to unlock the task
+   * @throws QueueException if an error occurred while performing the operation
+   */
+  public void unlock() throws QueueException, InterruptedException;
+
+  /**
+   * Reports the successful completion of this task. The task will automatically be unlocked and
+   * removed from the associated queue.
+   *
+   * @throws IllegalStateException if the current thread/worker doesn't own the lock to the task or
+   *     if the task is not currently or no longer locked
+   * @throws InterruptedException if the thread was interrupted while attempting to unlock the task
+   * @throws QueueException if an error occurred while performing the operation
+   */
+  public void complete() throws QueueException, InterruptedException;
+
+  /**
+   * Reports the failed completion of this task. The task will automatically be unlocked and removed
+   * from the associated queue. Such a task will not be re-queued for another attempt at being
+   * processed.
+   *
+   * @throws IllegalStateException if the current thread/worker doesn't own the lock to the task or
+   *     if the task is not currently or no longer locked
+   * @throws InterruptedException if the thread was interrupted while attempting to unlock the task
+   * @throws QueueException if an error occurred while performing the operation
+   */
+  public void fail() throws QueueException, InterruptedException;
+
+  /**
+   * Re-queues this task with the specified priority level into the associated queue for another
+   * attempt at being processed by re-inserting it into this queue if it is possible to do so
+   * immediately without violating capacity restrictions. This will typically be called by a worker
+   * after a failure occurs which permits another attempt at processing the task.
+   *
+   * @param level the new level for the new re-queued task
+   * @return <code>true</code> if the task was removed and re-added to this queue (only then will
+   *     the task be unlocked); <code>false</code> otherwise (the task is still considered locked)
+   * @throws IllegalStateException if the current thread/worker doesn't own the lock to the task or
+   *     if the task is not currently or no longer locked
+   * @throws QueueException if an error occurred while performing the operation
+   */
+  public boolean offerForRetry(PriorityLevel level) throws QueueException;
```

Comment:
❓ It seems like with the way this works now, a Task can be offered back into the queue with the same priority, but it will lose its original position, whereas if we only removed on success, a failure would keep its original position. Is losing the original position acceptable?
❓ Should a Worker be allowed to re-queue a Task at any priority? What happens when a priority service exists later and the priority it sets is violated by a worker?

Reply:
These were changed after an offline discussions.